File: PHP/DesignPatterns/Strategy.php
<?php

interface iTradingStrategy
{
    public function MakeTrades();
}

class RiskyTradingStrategy implements iTradingStrategy
{
    public function MakeTrades()
    {
        echo "Making risky trades!";
    }
}

class ModerateTradingStrategy implements iTradingStrategy
{
    public function MakeTrades()
    {
        echo "Making moderate trades.";
    }
}

class ConservativeTradingStrategy implements iTradingStrategy
{
    public function MakeTrades()
    {
        echo "Making safe trades.";
    }
}

class TradingFirm
{
    // Squirrel away the strategies:
    private $riskyStrategy;
    private $moderateStrategy;
    private $safeStrategy;

    private $currentStrategy;

    function __construct()
    {
        $this->riskyStrategy = new RiskyTradingStrategy();
        $this->moderateStrategy = new ModerateTradingStrategy();
        $this->safeStrategy = new ConservativeTradingStrategy();

        $this->currentStrategy = $this->moderateStrategy;
    }

    public function MarketUpdate($tradeConditions)
    {
        // Select the strategy based on market conditions:
        if($tradeConditions === "BearMarket")
        {
            $this->currentStrategy = $this->safeStrategy;
        }
        else if($tradeConditions  === "BullMarket")
        {
            $this->currentStrategy = $this->riskyStrategy;
        }
        else if($tradeConditions === "RecoveringMarket")
        {
            $this->currentStrategy = $this->moderateStrategy;
        }

        // Make trades with that strategy:
        $this->currentStrategy->MakeTrades();
    }
}

File: Java/DesignPatterns/Strategy.java
//Strategy demo

public interface TradingStrategy
{
 void MakeTrades();
}

public class RiskyTradingStrategy implements TradingStrategy
{
 public void MakeTrades()
 {
     System.out.println("Making risky trades!");
 }
}

public class ModerateTradingStrategy implements TradingStrategy
{
 public void MakeTrades()
 {
     System.out.println("Making moderate trades.");
 }
}

public class ConservativeTradingStrategy implements TradingStrategy
{
 public void MakeTrades()
 {
     System.out.println("Making safe trades.");
 }
}

//Important for getting the current state of the market:
public enum TradeConditions 
{
 BearMarket,
 BullMarket,
 RecoveringMarket
}

public class TradingFirm
{
 // Squirrel away the strategies to conserve GC:
 private TradingStrategy riskyStrategy;
 private TradingStrategy moderateStrategy;
 private TradingStrategy safeStrategy;

 // The currently used strategy:
 private TradingStrategy currentStrategy;

 public TradingFirm()
 {
     riskyStrategy = new RiskyTradingStrategy();
     moderateStrategy = new ModerateTradingStrategy();
     safeStrategy = new ConservativeTradingStrategy();

     currentStrategy = moderateStrategy;
 }

 public void MarketUpdate(TradeConditions marketCondition)
 {
     // Select the best strategy based on the market conditions:
     switch (marketCondition) {
         case BearMarket:
             currentStrategy = safeStrategy;
             break;
         case BullMarket:
             currentStrategy = riskyStrategy;
             break;
         case RecoveringMarket:
             currentStrategy = moderateStrategy;
             break;
         default:
             break;
     }

     // Make trades with that strategy:
     currentStrategy.MakeTrades();
 }

}